src/format_rules.ml

```diff
@@ -22,19 +22,42 @@ let add_diff sctx loc alias ~dir input output =
        ~targets:[]
        action)
 
+let rec subdirs_until_root dir =
+  match Path.parent dir with
+  | None -> [dir]
+  | Some d -> dir :: subdirs_until_root d
+
+let depend_on_existing_paths paths =
+  let open Build.O in
+  let build_id = Build.arr (fun x -> x) in
+  List.fold_left
+    ~f:(fun acc path ->
+      Build.if_file_exists
+        path
+        ~then_:(Build.path path)
+        ~else_:build_id
+      >>>
+      acc)
+    ~init:build_id
+    paths
+
+let depend_on_files ~named dir =
+  subdirs_until_root dir
+  |> List.map ~f:(fun dir -> Path.relative dir named)
+  |> depend_on_existing_paths
+
 let gen_rules sctx (config : Dune_file.Auto_format.t) ~dir =
   let loc = config.loc in
-  let files =
-    File_tree.files_of
-      (Super_context.file_tree sctx)
-      (Path.drop_build_context_exn dir)
-  in
+  let source_dir = Path.drop_build_context_exn dir in
+  let files = File_tree.files_of (Super_context.file_tree sctx) source_dir in
   let subdir = ".formatted" in
   let output_dir = Path.relative dir subdir in
   let alias = Build_system.Alias.make "fmt" ~dir in
   let alias_formatted = Build_system.Alias.make "fmt" ~dir:output_dir in
   let resolve_program = Super_context.resolve_program sctx ~loc:(Some loc) in
-  let setup_formatting file (arrows_acc, extra_deps_acc) =
+  let ocamlformat_deps = depend_on_files ~named:".ocamlformat" source_dir in
```

Comment:
I would put this under a lazy, just so that we don't do extra work if the directory doesn't contain source files.